File: backend/app/schemas/paper_trading.py
"""历史模拟交易 Pydantic Schemas。"""

from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, field_validator


# ---------- 会话 ----------

class CreateSessionRequest(BaseModel):
    start_date: date
    initial_cash: float
    name: Optional[str] = None

    @field_validator("initial_cash")
    @classmethod
    def validate_cash(cls, v: float) -> float:
        if v < 1000:
            raise ValueError("初始资金不能低于 1000 元")
        return v


class PositionSummary(BaseModel):
    """持仓聚合展示（按股票聚合多批次，展示加权均价）。"""
    stock_code: str
    stock_name: Optional[str]
    total_quantity: int
    avg_cost_price: float
    current_price: Optional[float]   # phase=open 时为开盘价，phase=close 时为收盘价
    market_value: Optional[float]
    profit_loss: Optional[float]
    profit_loss_pct: Optional[float]
    can_sell_quantity: int           # 排除当日买入批次后的可卖数量


class ClosedStockSummary(BaseModel):
    """已清仓股票：当前会话内对该代码已无 holding 批次，但存在 closed 批次（曾全部卖出）。"""
    stock_code: str
    stock_name: Optional[str]
    closed_batch_count: int
    # 本会话内该代码已实现盈亏：卖出净入金 − 买入净出金（均含手续费），比例 = 盈亏 / 买入总成本
    realized_profit_loss: float
    realized_profit_loss_pct: float


class SessionResponse(BaseModel):
    session_id: str
    name: Optional[str]
    start_date: date
    current_date: date
    current_phase: str               # open / close
    initial_cash: float
    available_cash: float
    status: str
    positions: list[PositionSummary] = []
    closed_stocks: list[ClosedStockSummary] = []
    total_asset: float
    total_profit_loss: float
    total_profit_loss_pct: float
    created_at: datetime
    # 展示用：对应「当前模拟日」的上一交易日收盘后的大盘温度（非模拟当日，避免未收盘当日被误用）
    market_temp_ref_date: Optional[date] = None
    market_temp_score: Optional[float] = None
    market_temp_level: Optional[str] = None


class SessionListItem(BaseModel):
    session_id: str
    name: Optional[str]
    start_date: date
    current_date: date
    current_phase: str
    initial_cash: float
    available_cash: float
    total_asset: float
    status: str
    created_at: datetime


class SessionListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[SessionListItem]


# ---------- 推进到收盘 / 下一交易日 ----------

class PhaseResponse(BaseModel):
    current_date: date
    current_phase: str
    available_cash: float
    positions: list[PositionSummary]
    closed_stocks: list[ClosedStockSummary] = []
    market_temp_ref_date: Optional[date] = None
    market_temp_score: Optional[float] = None
    market_temp_level: Optional[str] = None


class NextDayResponse(BaseModel):
    previous_date: date
    current_date: date
    current_phase: str
    available_cash: float
    positions: list[PositionSummary]
    closed_stocks: list[ClosedStockSummary] = []
    market_temp_ref_date: Optional[date] = None
    market_temp_score: Optional[float] = None
    market_temp_level: Optional[str] = None


class EndSessionResponse(BaseModel):
    """结束会话响应。"""
    session_id: str
    status: str


# ---------- 买入 / 卖出 ----------

class BuyRequest(BaseModel):
    stock_code: str
    price: float
    quantity: int

    @field_validator("quantity")
    @classmethod
    def validate_quantity(cls, v: int) -> int:
        if v <= 0 or v % 100 != 0:
            raise ValueError("买入数量必须为 100 的正整数倍")
        return v


class SellRequest(BaseModel):
    stock_code: str
    price: float
    quantity: int

    @field_validator("quantity")
    @classmethod
    def validate_quantity(cls, v: int) -> int:
        if v <= 0 or v % 100 != 0:
            raise ValueError("卖出数量必须为 100 的正整数倍")
        return v


class OrderResponse(BaseModel):
    order_id: int
    order_type: str
    stock_code: str
    stock_name: Optional[str]
    trade_date: date
    price: float
    quantity: int
    amount: float
    commission: float
    cash_after: float
    # 订单写入数据库的时间，用于同日内多笔成交的先后展示
    created_at: datetime


class OrderListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[OrderResponse]


# ---------- 图表数据 ----------

class ChartBar(BaseModel):
    date: str
    open: Optional[float]
    high: Optional[float]   # phase=open 时日 K 最新一根为 None
    low: Optional[float]    # phase=open 时日 K 最新一根为 None
    close: Optional[float]  # phase=open 时日 K 最新一根为 None
    volume: Optional[float] # phase=open 时日 K 最新一根为 None（避免提前暴露全日量）
    prev_close: Optional[float]
    pct_change: Optional[float]
    ma5: Optional[float]
    ma10: Optional[float]
    ma20: Optional[float]
    ma60: Optional[float]
    macd_dif: Optional[float]  # phase=open 时最新一根为 None
    macd_dea: Optional[float]  # phase=open 时最新一根为 None
    macd_hist: Optional[float] # phase=open 时最新一根为 None


class ChartDataResponse(BaseModel):
    stock_code: str
    stock_name: Optional[str]
    period: str
    open_price: Optional[float]   # 当日开盘价（快捷填充用）
    close_price: Optional[float]  # 当日收盘价（phase=close 时有值）
    limit_up: Optional[float]     # 涨停价
    limit_down: Optional[float]   # 跌停价
    data: list[ChartBar]


# ---------- 推荐 / 筛选 ----------

class StockQuote(BaseModel):
    stock_code: str
    stock_name: Optional[str]
    open: Optional[float]
    close: Optional[float]    # phase=open 时为 None
    pct_change: Optional[float]  # phase=open 时为 None
    volume: Optional[float]
    limit_up: Optional[float]
    limit_down: Optional[float]


class RecommendResponse(BaseModel):
    trade_date: date
    phase: str
    items: list[StockQuote]


class ScreenResponse(BaseModel):
    trade_date: date
    total: int
    page: int
    page_size: int
    items: list[StockQuote]


# ---------- 交易日列表 ----------

class TradingDatesResponse(BaseModel):
    dates: list[str]
    min_date: str
    max_date: str


# ---------- 股票解析（代码 / 名称模糊）与资料快照 ----------

class StockResolveItem(BaseModel):
    stock_code: str
    stock_name: Optional[str]
    instrument_kind: Literal["stock", "index"] = "stock"


class StockResolveResponse(BaseModel):
    items: list[StockResolveItem]


class StockInfoBasicBlock(BaseModel):
    stock_code: str
    stock_name: Optional[str]
    exchange: Optional[str]
    market: Optional[str]
    industry_name: Optional[str]
    region: Optional[str]
    list_date: Optional[str]


class StockInfoDailyBlock(BaseModel):
    """当前模拟日对应 stock_daily_bar 一行；phase=open 时不返回未揭晓的高、低、收、涨跌幅。"""
    trade_date: str
    open: Optional[float]
    high: Optional[float]
    low: Optional[float]
    close: Optional[float]
    prev_close: Optional[float]
    pct_change: Optional[float]
    volume: Optional[float]
    amount: Optional[float]
    amplitude: Optional[float]
    turnover_rate: Optional[float]
    pe_ttm: Optional[float]
    pb: Optional[float]
    total_market_cap: Optional[float]
    float_market_cap: Optional[float]


class StockInfoFinancialBlock(BaseModel):
    """截至 end_date（含）最近一期财报行，字段来自 stock_financial_report。"""
    report_date: str
    report_type: Optional[str]
    roe: Optional[float]
    roe_dt: Optional[float]
    debt_to_assets: Optional[float]
    roa: Optional[float]
    gross_margin: Optional[float]
    net_margin: Optional[float]
    revenue: Optional[float]
    net_profit: Optional[float]
    eps: Optional[float]
    bps: Optional[float]


class PaperStockInfoResponse(BaseModel):
    stock_code: str
    end_date: str
    phase: str
    basic: StockInfoBasicBlock
    daily: Optional[StockInfoDailyBlock]
    financial: Optional[StockInfoFinancialBlock]

File: frontend/src/api/paperTrading.ts
import http from './http'

// ---------- Types ----------

export interface CreateSessionRequest {
  start_date: string
  initial_cash: number
  name?: string
}

export interface PositionSummary {
  stock_code: string
  stock_name: string | null
  total_quantity: number
  avg_cost_price: number
  current_price: number | null
  market_value: number | null
  profit_loss: number | null
  profit_loss_pct: number | null
  can_sell_quantity: number
}

/** 已清仓：当前对该股已无 holding 批次，曾产生过 closed 批次 */
export interface ClosedStockSummary {
  stock_code: string
  stock_name: string | null
  closed_batch_count: number
  /** 本会话该代码已实现盈亏（元），卖出净入 − 买入净出（含手续费） */
  realized_profit_loss: number
  /** 相对买入总成本的比例，如 0.052 表示 5.2% */
  realized_profit_loss_pct: number
}

export interface SessionResponse {
  session_id: string
  name: string | null
  start_date: string
  current_date: string
  current_phase: 'open' | 'close'
  initial_cash: number
  available_cash: number
  status: string
  positions: PositionSummary[]
  closed_stocks: ClosedStockSummary[]
  total_asset: number
  total_profit_loss: number
  total_profit_loss_pct: number
  created_at: string
  /** 大盘温度对应的交易日：当前模拟日的上一交易日（非模拟当日） */
  market_temp_ref_date: string | null
  market_temp_score: number | null
  market_temp_level: string | null
}

export interface SessionListItem {
  session_id: string
  name: string | null
  start_date: string
  current_date: string
  current_phase: 'open' | 'close'
  initial_cash: number
  available_cash: number
  total_asset: number
  status: string
  created_at: string
}

export interface SessionListResponse {
  total: number
  page: number
  page_size: number
  items: SessionListItem[]
}

export interface PhaseResponse {
  current_date: string
  current_phase: 'open' | 'close'
  available_cash: number
  positions: PositionSummary[]
  closed_stocks: ClosedStockSummary[]
  market_temp_ref_date: string | null
  market_temp_score: number | null
  market_temp_level: string | null
}

export interface NextDayResponse {
  previous_date: string
  current_date: string
  current_phase: 'open' | 'close'
  available_cash: number
  positions: PositionSummary[]
  closed_stocks: ClosedStockSummary[]
  market_temp_ref_date: string | null
  market_temp_score: number | null
  market_temp_level: string | null
}

export interface OrderResponse {
  order_id: number
  order_type: 'buy' | 'sell'
  stock_code: string
  stock_name: string | null
  trade_date: string
  price: number
  quantity: number
  amount: number
  commission: number
  cash_after: number
  /** 订单落库时间（ISO），同一天多笔时区分先后 */
  created_at?: string
}

export interface OrderListResponse {
  total: number
  page: number
  page_size: number
  items: OrderResponse[]
}

export interface ChartBar {
  date: string
  open: number | null
  high: number | null
  low: number | null
  close: number | null
  volume: number | null
  prev_close: number | null
  pct_change: number | null
  ma5: number | null
  ma10: number | null
  ma20: number | null
  ma60: number | null
  macd_dif: number | null
  macd_dea: number | null
  macd_hist: number | null
}

export interface ChartDataResponse {
  stock_code: string
  stock_name: string | null
  period: string
  open_price: number | null
  close_price: number | null
  limit_up: number | null
  limit_down: number | null
  data: ChartBar[]
}

export interface StockQuote {
  stock_code: string
  stock_name: string | null
  open: number | null
  close: number | null
  pct_change: number | null
  volume: number | null
  limit_up: number | null
  limit_down: number | null
}

export interface RecommendResponse {
  trade_date: string
  phase: string
  items: StockQuote[]
}

export interface ScreenResponse {
  trade_date: string
  total: number
  page: number
  page_size: number
  items: StockQuote[]
}

export interface TradingDatesResponse {
  dates: string[]
  min_date: string
  max_date: string
}

export interface StockResolveItem {
  stock_code: string
  stock_name: string | null
  /** 后端：stock=个股，index=指数点位标的 */
  instrument_kind?: 'stock' | 'index'
}

export interface StockResolveResponse {
  items: StockResolveItem[]
}

export interface StockInfoBasicBlock {
  stock_code: string
  stock_name: string | null
  exchange: string | null
  market: string | null
  industry_name: string | null
  region: string | null
  list_date: string | null
}

export interface StockInfoDailyBlock {
  trade_date: string
  open: number | null
  high: number | null
  low: number | null
  close: number | null
  prev_close: number | null
  pct_change: number | null
  volume: number | null
  amount: number | null
  amplitude: number | null
  turnover_rate: number | null
  pe_ttm: number | null
  pb: number | null
  total_market_cap: number | null
  float_market_cap: number | null
}

export interface StockInfoFinancialBlock {
  report_date: string
  report_type: string | null
  roe: number | null
  roe_dt: number | null
  debt_to_assets: number | null
  roa: number | null
  gross_margin: number | null
  net_margin: number | null
  revenue: number | null
  net_profit: number | null
  eps: number | null
  bps: number | null
}

export interface PaperStockInfoResponse {
  stock_code: string
  end_date: string
  phase: string
  basic: StockInfoBasicBlock
  daily: StockInfoDailyBlock | null
  financial: StockInfoFinancialBlock | null
}

export interface ScreenParams {
  trade_date: string
  pct_change_min?: number
  pct_change_max?: number
  volume_min?: number
  volume_max?: number
  ma_golden_cross?: string
  macd_golden_cross?: boolean
  page?: number
  page_size?: number
}

// ---------- API ----------

export const paperTradingApi = {
  createSession: (data: CreateSessionRequest) =>
    http.post<SessionResponse>('/paper-trading/sessions', data),

  listSessions: (params?: { status?: string; page?: number; page_size?: number }) =>
    http.get<SessionListResponse>('/paper-trading/sessions', { params }),

  getSession: (sessionId: string) =>
    http.get<SessionResponse>(`/paper-trading/sessions/${sessionId}`),

  advanceToClose: (sessionId: string) =>
    http.post<PhaseResponse>(`/paper-trading/sessions/${sessionId}/advance-to-close`),

  nextDay: (sessionId: string) =>
    http.post<NextDayResponse>(`/paper-trading/sessions/${sessionId}/next-day`),

  endSession: (sessionId: string) =>
    http.post(`/paper-trading/sessions/${sessionId}/end`),

  buy: (sessionId: string, data: { stock_code: string; price: number; quantity: number }) =>
    http.post<OrderResponse>(`/paper-trading/sessions/${sessionId}/buy`, data),

  sell: (sessionId: string, data: { stock_code: string; price: number; quantity: number }) =>
    http.post<OrderResponse>(`/paper-trading/sessions/${sessionId}/sell`, data),

  listOrders: (
    sessionId: string,
    params?: {
      order_type?: string
      stock_code?: string
      page?: number
      page_size?: number
      sort?: 'asc' | 'desc'
    }
  ) => http.get<OrderListResponse>(`/paper-trading/sessions/${sessionId}/orders`, { params }),

  getChartData: (params: {
    stock_code: string
    end_date: string
    phase: string
    period?: string
    /** false 时配合 limit 仅取最近 N 根；默认 true 从库中最早一根取至 end_date */
    full_history?: boolean
    limit?: number
  }) => http.get<ChartDataResponse>('/paper-trading/chart-data', { params }),

  resolveStock: (params: { q: string; limit?: number }) =>
    http.get<StockResolveResponse>('/paper-trading/resolve-stock', { params }),

  getStockInfo: (params: { stock_code: string; end_date: string; phase: string }) =>
    http.get<PaperStockInfoResponse>('/paper-trading/stock-info', { params }),

  recommend: (params: { trade_date: string; phase: string; count?: number }) =>
    http.get<RecommendResponse>('/paper-trading/recommend', { params }),

  screen: (params: ScreenParams) =>
    http.get<ScreenResponse>('/paper-trading/screen', { params }),

  getTradingDates: (params: { start: string; end: string }) =>
    http.get<TradingDatesResponse>('/paper-trading/trading-dates', { params }),
}
